akka-stream/src/main/scala/akka/stream/impl/io/TLSActor.scala

```diff
@@ -165,6 +165,22 @@ private[akka] class TLSActor(settings: ActorMaterializerSettings,
     params.setServerNames(Collections.singletonList(serverName))
   }
 
+  private def cloneSSLParameters(parameters: SSLParameters): SSLParameters = {
+    val clone = new SSLParameters()
+
+    clone.setCipherSuites(parameters.getCipherSuites)
+    clone.setProtocols(parameters.getProtocols)
+    clone.setWantClientAuth(parameters.getWantClientAuth)
+    clone.setNeedClientAuth(parameters.getNeedClientAuth)
+    clone.setEndpointIdentificationAlgorithm(parameters.getEndpointIdentificationAlgorithm)
+    clone.setAlgorithmConstraints(parameters.getAlgorithmConstraints)
+    clone.setServerNames(parameters.getServerNames)
+    clone.setSNIMatchers(parameters.getSNIMatchers)
+    clone.setUseCipherSuitesOrder(parameters.getUseCipherSuitesOrder)
+
+    clone
+  }
```

Comment:
This can be done in one go without cloning the params if `hostInfo` is `None`:

```scala
private def applySNI: Unit = for {
  params <- params.sslParameters
  (hostname, _) <- hostInfo
} yield {
  //first copy the mutable SLLParameters before modifying to prevent race condition in `setServerNames`
  val clone = new SSLParameters()
  clone.setCipherSuites(params.getCipherSuites)
  clone.setProtocols(params.getProtocols)
  clone.setWantClientAuth(params.getWantClientAuth)
  clone.setNeedClientAuth(params.getNeedClientAuth)
  clone.setEndpointIdentificationAlgorithm(params.getEndpointIdentificationAlgorithm)
  clone.setAlgorithmConstraints(params.getAlgorithmConstraints)
  clone.setSNIMatchers(params.getSNIMatchers)
  clone.setUseCipherSuitesOrder(params.getUseCipherSuitesOrder)

  // apply the changes
  clone.setServerNames(Collections.singletonList(new SNIHostName(hostname)))
  engine.setSSLParameters(parameters)
}
```